enhancements/sig-architecture/224-event-based-manifestwork/README.md

```diff
@@ -0,0 +1,1042 @@
+`ManifestWork` on the hub
+
+```golang
+mwCloudEventClient := NewCloudEventSourceClient[*workv1.ManifestWork](
+			"hub1-manifestworkreplicaset-controller",
+			mqttCloudEventSubClient, mqttCloudEventPubClient,
+			manifestWorkLister,
+			manifestWorkStatusHashGetter,
+		)
+
+mwCloudEventClient.AddCodec("io.open-cluster-management.works.v1alpha1.manifestbundle", manifestbundleCodec)
+```
+
+Also the developers can easily to build an agent client for `ManifestWork` on the managed cluster
+```golang
+mwCloudEventClient := NewCloudEventSourceClient[*workv1.ManifestWork](
+			"cluster1-work-agent",
+      "cluster1",
+			mqttCloudEventSubClient, mqttCloudEventPubClient,
+			manifestWorkLister,
+			manifestWorkStatusHashGetter,
+		)
+
+// add multiple encoders/decoders to support different cloud event payload
+mwCloudEventClient.AddCodec("io.open-cluster-management.works.v1alpha1.manifestbundle", manifestbundleCodec)
+mwCloudEventClient.AddCodec("io.open-cluster-management.works.v1alpha1.manifest", manifestCodec)
+```
+
+By default, we provide a work client builder for developers, developers can enable the cloud event based on the MQTT
+for `ManifestWork` with MQTT options, the builder returns a work client holder, we can get the work client set from this
+holder, e.g.
+
+```golang
+workClientHolder, err := workclient.NewWorkClientBuilder("work-agent-id").
+		WithHubKubeConfig(hubKubeConfig).
+		WithMQTTOptions(o.mqttOptions).
+		WithClusterName(o.SpokeClusterName).
+		WithCloudEventTypes(workclient.ManifestBundle, workclient.Manifest).
+		NewWorkClientHolder(ctx)
+if err != nil {
+		return err
+}
+
+workClientSet := hubWorkClientHolder.GetWorkClientSet()
+```
+
+Using the work client set, developers can access the ManifestWork or create a Kubernetes informer and lister. This
+process closely mirrors what developers typically do in a Kubernetes environment.
+
+```golang
+workInformerFactory := workinformers.NewSharedInformerFactoryWithOptions(workClientSet, 5*time.Minute, workinformers.WithNamespace(o.SpokeClusterName))
+manifestWorkClient := workClientSet.WorkV1().ManifestWorks(o.SpokeClusterName)
+manifestWorkInformer := workInformerFactory.Work().V1().ManifestWorks()
+manifestWorkLister := workInformerFactory.Work().V1().ManifestWorks().Lister().ManifestWorks(o.SpokeClusterName)
+```
+
+### Failure Handling
+
+To ensure the integrity of the event-based manifestworks system during source or work agent restart or reconnection, the following optimized resync mechanism is employed:
+
+1. On start, the source subscribes to the spec resync event topic `/sources/clusters/+/specresync` from the broker.
+2. On start, the work agent subscribes to the status resync event topic `/sources/+/clusters/statusresync` from the broker.
+3. Upon work agent restarts or reconnects, the work agent generates a spec resync event with resource versions and sends it to the broker.
```

Comment:
How about status events that happened while the broker is offline or not reachable? Those would be lost?

I'm thinking in the policy case, we want to record every single compliance event which is why #98 proposed to have a local queue cache of compliance events until they can be recorded in the centralized place.

Reply:
This depends on message QoS?

Reply:
By default, we will set QoS to 1 to make sure the event is delivered to the broker, but if the broker is offline/unreachable, I think we need to have a retry for the agent to send the event when the broker becomes reachable again, so a local cache may be required for the policy case

Reply:
@mprahl MQTT Agent (or our agent) can be in a ship where it cannot connect to the MQTT Broker for an extended period. If for that agent, it is important to replay history, this has to stored by the agent. Not all agents will need this though. Does th make sense @skeeey @mprahl